src/main/java/org/jivesoftware/site/PluginManager.java

```diff
@@ -0,0 +1,502 @@
+    {
+        final String repoPath;
+        switch ( parsedRequest.product )
+        {
+            case openfire_plugin:
+                switch ( parsedRequest.releaseTrack )
+                {
+                    case BETA_RELEASE:
+                        repoPath = openfirePluginsBetaPath;
+                        break;
+
+                    case DEV_RELEASE:
+                        repoPath = openfirePluginsDevPath;
+                        break;
+
+                    case RELEASE:
+                    default:
+                        repoPath = openfirePluginsPath;
+                        break;
+                }
+                break;
+
+            case spark_plugin:
+                repoPath = sparkPluginsPath;
+                break;
+
+            default:
+                return null;
+        }
+
+        final Collection<Metadata> plugins = getPlugins( Paths.get( repoPath ) );
+
+        Stream<Metadata> stream = plugins.stream();
+        if ( parsedRequest.pluginVersion != null && parsedRequest.pluginVersion.toUpperCase().endsWith( "-SNAPSHOT" ) )
+        {
+            stream = stream.filter( metadata -> metadata.isSnapshot );
+        }
+        else
+        {
+            stream = stream.filter( metadata -> metadata.isRelease );
+        }
+
+        stream = stream.filter( metadata -> (parsedRequest.pluginName + ".jar").equalsIgnoreCase( metadata.pluginFileName ) );
+
+        if ( parsedRequest.pluginVersion != null )
+        {
+            stream = stream.filter( metadata -> metadata.version.equalsIgnoreCase( parsedRequest.pluginVersion ) );
+            stream = stream.sorted( Comparator.comparing( metadata -> metadata.mavenFile.toFile().lastModified() ) );
+        }
+        else
+        {
+            // FIXME: don't use alphabetical sorting
+            stream = stream.sorted( Comparator.comparing( metadata -> metadata.version ) );
+        }
+
+        final Optional<Metadata> first = stream.findFirst();
+
+        return first.map( metadata -> metadata.mavenFile ).orElse( null );
+    }
+
+    /**
+     * Retrieves all metadata for every plugin found in the path.
+     *
+     * This method caches results. This cache is used to answer subsequent requests to this method.
+     * When a cache entry is stale, a refresh is computed. During the computation, the stale result
+     * is returned.
+     *
+     * @param path The path to index.
+     * @return All metadata on the path.
+     */
+    private static Set<Metadata> getPlugins( Path path ) throws IOException
+    {
+        synchronized ( cache )
+        {
+            final Set<Metadata> cachedResult = cache.get( path );
+            final boolean isRefreshing = refreshing.computeIfAbsent( path, p -> false );
+            final long refreshed = lastRefreshed.computeIfAbsent( path, p -> System.currentTimeMillis() );
+            final boolean isStale = System.currentTimeMillis() - refreshed > 10*60*1000;
```

Comment:
I'd like to see `10*60*1000` replaced with something like `Duration.ofMinutes(10).toMillis()` as it's far more explicit as to the cache duration.